### PR TITLE
SHA512 based JSONB digest function

### PR DESCRIPTION
**Pull request by sumerman**
A custom SHA512 based JSONB digest.

Save for hash collisions, for any pair of `j1` and `j2` the statement `SELECT jsonb_digest(j1) = jsonb_digest(j2)` is true iff `j1 == j2` are _semantically_ equivalent.

There is a couple of alternatives, unfortunately neither should be used in a UNIQUE index.

`pg_catalog.sha512(json_value::text::bytea)` has two drawbacks:
- There isn't enough control over `jsonb` -> `text` conversion.
  Although, the existing implementation relies on the same mechanism
  as `jsonb_digest` and should always keep object keys in the fixed
  order, there's no way to explicitly set specifics of its behaviour
  like indentation or white-space padding.
- JSONB stores all numeric values in PostgreSQL `numeric` type.
  It's not just precise, it accurately preserves trailing fractional zeroes.
  In other words, values `1.01` and `1.010` will be represented differently,
  despite being equal.
 
There's also `jsonb_hash_extended` that addresses the shortcomings, described above. Unfortunately, it is limited to 64 bit: good enough for hash tables, probably a bad idea for a UNIQUE index. This implementation is largely based on `jsonb_hash_extended`.

A prerequisite for #130

### CODE REVIEW

**Review by JamesGuthrie — CHANGES_REQUESTED** [review]

File: src/raw.rs [hunk 1]
```diff
@@ -59,4 +59,8 @@ macro_rules! raw_type {
 #[derive(Clone, Copy)]
 pub struct bytea(pub pg_sys::Datum);
 
+#[derive(Clone, Copy)]
+pub struct Jsonb(pub pg_sys::Datum);
+
 raw_type!(bytea, pg_sys::BYTEAOID, pg_sys::BYTEAARRAYOID);
+raw_type!(Jsonb, pg_sys::JSONBOID, pg_sys::JSONBARRAYOID);
```
[the remark this answers is not in the document]
Comment (JamesGuthrie) [reply src/raw.rs:66]: In case it wasn't clear: I don't think that we should change anything in how we've implemented this.

File: src/jsonb_digest.rs [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    use pg_sys::*;
+    use sha2::{Digest, Sha512};
+
+    if jsonb_datum.0 == 0 {
+        panic!("This is a strict function. NULL argument is not expected.");
+    }
+
+    let mut hasher = Sha512::new();
+
+    let jsonb = pg_detoast_datum(jsonb_datum.0 as *mut varlena) as *mut Jsonb;
```
[the remark this answers is not in the document]
Comment (sumerman) [reply src/jsonb_digest.rs:15]: Yes and no. Everything within a memory context will be freed at once. Usually at the end of the query in this case. In the upcoming commit I've separated json iterator into its own thing and handling this explicitly.

File: src/jsonb_digest.rs [hunk 3]
```diff
@@ -0,0 +1,118 @@
+use pgx::*;
+
+#[allow(non_upper_case_globals)]
+#[pg_extern(immutable, strict, parallel_safe)]
+pub unsafe fn jsonb_digest(jsonb_datum: crate::raw::Jsonb) -> Vec<u8> {
+    use pg_sys::*;
+    use sha2::{Digest, Sha512};
+
+    if jsonb_datum.0 == 0 {
```
[the remark this answers is not in the document]
Comment (sumerman) [reply src/jsonb_digest.rs:9]: Good to know! Thanks!

File: src/jsonb_digest.rs [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+                ),
+            },
+            _ => elog(
+                PgLogLevel::ERROR,
+                format!("invalid JsonbIteratorNext rc: {}", r).as_str(),
+            ),
+        }
+    }
+
+    Vec::from(hasher.finalize().as_slice())
```
[the remark this answers is not in the document]
Comment (sumerman) [reply src/jsonb_digest.rs:69]: Good point. But we if chase this, I'll be cutting it too close to GA. @cevian WDYT?

File: src/raw.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (sumerman) [reply src/raw.rs:66]: `pgx::JsonB` converts the whole thing into a string and then parses it. Not what I wanted. Built my own type already.

File: src/jsonb_digest.rs [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+                    let num_str_ptr = numeric_normalize(jsonb_val.val.numeric);
+                    let num_str = std::ffi::CStr::from_ptr(num_str_ptr).to_str().unwrap();
+                    hasher.update(num_str.as_bytes());
+                    pfree(num_str_ptr as *mut std::os::raw::c_void);
+                }
+                jbvType_jbvBool => {
+                    let bytes = if jsonb_val.val.boolean {
+                        0x02u8.to_le_bytes()
+                    } else {
+                        0x04u8.to_le_bytes()
```
[the remark this answers is not in the document]
Comment (sumerman) [reply src/jsonb_digest.rs:53]: Parroting from `jsonb_hash_extended`, doesn't really matter AFAICT, just needs to be distinct from everything else.

File: src/jsonb_digest.rs [hunk 6]
```diff
@@ -0,0 +1,118 @@
+use pgx::*;
+
+#[allow(non_upper_case_globals)]
+#[pg_extern(immutable, strict, parallel_safe)]
+pub unsafe fn jsonb_digest(jsonb_datum: crate::raw::Jsonb) -> Vec<u8> {
```
[the remark this answers is not in the document]
Comment (sumerman) [reply src/jsonb_digest.rs:5]: As I've mentioned elsewhere, I've separated JSON traversal into an abstraction of its own. All unsafe bits will be hidden behind.

File: src/jsonb_digest.rs [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+    use pg_sys::*;
+    use sha2::{Digest, Sha512};
```
[the remark this answers is not in the document]
Comment (sumerman) [reply src/jsonb_digest.rs:7]: No particular reason. Changed that already.

File: src/jsonb_digest.rs [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+        );
+        match r {
+            JsonbIteratorToken_WJB_DONE => break,
+            JsonbIteratorToken_WJB_BEGIN_ARRAY => hasher.update(pg_sys::JB_FARRAY.to_le_bytes()),
+            JsonbIteratorToken_WJB_BEGIN_OBJECT => hasher.update(pg_sys::JB_FOBJECT.to_le_bytes()),
+            JsonbIteratorToken_WJB_END_ARRAY | JsonbIteratorToken_WJB_END_OBJECT => {}
+            JsonbIteratorToken_WJB_KEY
+            | JsonbIteratorToken_WJB_VALUE
+            | JsonbIteratorToken_WJB_ELEM => match jsonb_val.type_ as jbvType {
+                jbvType_jbvNull => hasher.update(0x01u8.to_le_bytes()),
```
[the remark this answers is not in the document]
Comment (sumerman) [reply src/jsonb_digest.rs:32]: Parroting from jsonb_hash_extended, doesn't really matter AFAICT, just needs to be distinct from everything else.

**Comment by sumerman** [comment]
Fun fact, `json_digest` is ever so slightly faster than `sha512(json::text::bytea)`

```
SELECT sha512(gen.json::text::bytea)
FROM (
	SELECT jsonb_object_agg(n::text, n) AS json
	FROM generate_series(1, 2000000) AS gs (n)
) AS gen;

 Planning Time: 0.065 ms
 Execution Time: 2306.121 ms

SELECT jsonb_digest(gen.json)
FROM (
	SELECT jsonb_object_agg(n::text, n) AS json
	FROM generate_series(1, 2000000) AS gs (n)
) AS gen;

 Planning Time: 0.062 ms
 Execution Time: 2195.345 ms
```

File: src/jsonb_digest.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (sumerman) [reply src/jsonb_digest.rs:66]: As discussed elsewhere `sha512` is actually 64 bytes long and `varlena` header end's up being a much smaller fraction 😅

File: src/iterable_jsonb.rs [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+                );
+                None
+            }
+        }
+    }
+}
+
+impl<'a> TokenIterator<'a> {
+    #[inline]
+    fn extract_string_value(jsonb_val: &mut JsonbValue) -> &'a str {
```
Comment (JamesGuthrie) [inline src/iterable_jsonb.rs:174]: nit: Here's a potential candidate for `unsafe fn`. It's only safe to call this when the caller has validated the `JsonbValue`'s type is 'jbvString'.

File: src/jsonb_digest.rs [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+            Bool(false) => hasher.update(0x03u8.to_le_bytes()),
+            String(str) => {
+                hasher.update(0x04u8.to_le_bytes());
+                hasher.update(str.as_bytes())
+            }
+            Token::Numeric(numeric) => {
+                hasher.update(0x05u8.to_le_bytes());
+                hasher.update(numeric.to_str().as_bytes())
+            }
+            //
```
[the remark this answers is not in the document]
Comment (sumerman) [reply src/jsonb_digest.rs:53]: No, I just wanted a visual separation between "scalars", "objects" and "arrays". Perhaps I should name them.

**Review by jgpruitt — COMMENTED** [review]
LGTM
